### PR TITLE
0.4 release doc

### PR DESCRIPTION
**Pull request by cnpetra**
This PR updates the user manual with the new features since 0.3 release.

### CODE REVIEW

File: doc/src_usermanual/techrep_main.tex [hunk 1]
```diff
@@ -617,7 +617,7 @@ \subsubsection{The C++ interface}
 \begin{itemize} 
 \item first six: see \texttt{eval\_cons}.
 \item \texttt{nnzJacS}, \texttt{iJacS}, \texttt{jJacS}, \texttt{MJacS} are for number of nonzeros, $(i,j)$ indexes, and nonzero values of   the sparse Jacobian.
-\item  JacD: dense Jacobian as a contiguous array storing the matrix by rows; the array is ``primed'' to support double indexing \texttt{JacD[i][j]}.
+\item \texttt{JacD} should contain the Jacobian with respect to the dense variables of the MDS problem. The array should store this Jacobian submatrix row-wise, meaning that the each row of the Jacobian is contiguous in memory and starts right after the previous row.
 \end{itemize}
 
 Notes for implementer of this method:
```
Comment (nychiang) [inline doc/src_usermanual/techrep_main.tex:623]: Should we use \warningcp{Note:}? like line 643

**Comment by nychiang** [comment]
Now [this note]( https://github.com/LLNL/hiop/blob/4056de05cb62d3479288abc2beda108197af7e2f/doc/src_usermanual/techrep_main.tex#L684) is not true any more.
Now lambda is not sorted, i.e., it uses the same order as how the constraints are defined by `get_cons_info`.
Reminder: I made this change due to the API of gollnlp

File: doc/src_usermanual/techrep_main.tex [hunk 2]
```diff
@@ -669,7 +669,7 @@ \subsubsection{The C++ interface}
 			      const double* lambda, bool new_lambda,
 			      const long long& nsparse, const long long& ndense, 
 			      const int& nnzHSS, int* iHSS, int* jHSS, double* MHSS, 
-			      double** HDD,
+			      double* HDD,
 			      int& nnzHSD, int* iHSD, int* jHSD, double* MHSD);
 \end{lstlisting}
```
Comment (nychiang) [inline doc/src_usermanual/techrep_main.tex:675]: I think it's better if we add another note to mention that only a triangular part is needed to pass into hiop

File: doc/src_usermanual/techrep_main.tex [hunk 3]
```diff
@@ -784,45 +784,210 @@ \section{Solver options}
 
 \warningcp{Note:} Options set in \texttt{hiop.options} file at runtime overwrite options set at runtime via the above API.
 
+\subsection{Termination criteria and output}
 
+\noindent \textbf{tolerance}: maximum (absolute)  NLP optimality error allowed at the optimal solution. Double values in $[10e-14,0.1]$. Default value: $10e-8$.
```
Comment (nychiang) [inline doc/src_usermanual/techrep_main.tex:789]: to keep consistence, should we use $10^{-14}$ and $10^{-8}$ here? See the next option `acceptable_tolerance` for example

File: doc/src_usermanual/techrep_main.tex [hunk 4]
```diff
@@ -784,45 +784,210 @@ \section{Solver options}
...
+\noindent \textbf{kappa2}: shift projection parameter used in initialization for doubly bounded variables. Double values in $[10^{-8},0.49999]$. Default value: $0.01$. 
+
+\medskip
+
+\noindent \textbf{smax}: the primal-dual IPM equations are rescaled when the average value of the is larger than this threshold value. Double values in $[1,10^7]$. Default value: $100$. 
+
+\medskip
+
+
+\noindent \textbf{duals\_init}: type of the update for the initialization of Lagrange multipliers corresponding to  the equality constraints. Possible values one of the the strings "lsq" (least-square (LSQ) solve initialization) and "zero"(multipliers are set identically to zero). Default value is "lsq".
```
Comment (nychiang) [inline doc/src_usermanual/techrep_main.tex:853]: change "zero" to \``zero" and "lsq" to \``lsq"

File: doc/src_usermanual/techrep_main.tex [hunk 5]
```diff
@@ -784,45 +784,210 @@ \section{Solver options}
...
+\noindent \textbf{secant\_memory\_len}: size of the memory (number of $(s,y)$ pairs) of the Hessian secant approximation. Integer values between $0$ and $256$. Default value $6$.
+
+\subsection{Linear algebra computational kernels}
+
+\noindent \textbf{KKTLinsys}: type of KKT linear system \textit{formulation} used internally
+\begin{itemize}
+\item ``auto'' (default): decided by \Hi based on the type of interface/NLP solved and ``compute\_mode`` and ``Hessian`` options
+\item ``xycyd'': symmetric indefinite (less stable but smaller size)
+\item ``xdycyd'': symmetric indefinite (more stable but  larger size)
+\item  ``full'': unsymetric suitable for LU solvers (experimental)
```
Comment (nychiang) [inline doc/src_usermanual/techrep_main.tex:897]: should we mention this option in this release?

File: doc/src_usermanual/techrep_main.tex [hunk 6]
```diff
@@ -784,45 +784,210 @@ \section{Solver options}
...
+\begin{itemize}
+\item ``none'' (default): perform no problem scaling 
+\item ``gradient'':  will scale the problem such that the inf-norm of gradient at the initial point is less or equal to the value of ``scaling\_max\_grad'' option.
+\end{itemize}
+
+\medskip
+
+\noindent \textbf{scaling\_max\_grad}: The user's NLP will be rescaled if the inf-norm of the gradient at the starting point is larger than the value of this option. Double values in $[10^{-20}, 10^{20}]$. Default value $100$.
+
+
```
Comment (nychiang) [inline doc/src_usermanual/techrep_main.tex:989]: how about the other options introduced from #197 ?

File: doc/src_usermanual/techrep_main.tex (hunk 5 above)
Comment (cnpetra) [reply doc/src_usermanual/techrep_main.tex:901]: it says is experimental ;) let's leave it there, we may forget to add it later on.

**Comment by cnpetra** [comment]
> Now [this note](https://github.com/LLNL/hiop/blob/4056de05cb62d3479288abc2beda108197af7e2f/doc/src_usermanual/techrep_main.tex#L684) is not true any more.
> Now lambda is not sorted, i.e., it uses the same order as how the constraints are defined by `get_cons_info`.
> Reminder: I made this change due to the API of gollnlp

good catch, as are the others... I've addressed all of them.

**Review by nychiang — APPROVED** [review]

File: doc/src_usermanual/techrep_main.tex (hunk 1 above)
Comment (cnpetra) [reply doc/src_usermanual/techrep_main.tex]: done

**Pull request merged**